filebeat/input/filestream/internal/task/group.go

```diff
@@ -0,0 +1,141 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package task
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/semaphore"
+)
+
+type Group struct {
+	sem *semaphore.Weighted
+	wg  *sync.WaitGroup
+
+	stopTimeout time.Duration
+	logErr      func(error)
+
+	ctx       context.Context
+	cancelCtx context.CancelFunc
+}
+
+type Logger interface {
+	Errorf(format string, args ...interface{})
+}
+
+// NewGroup returns a new task group which will run tasks on a goroutine. See
+// Group.Go for details.
+//
+// The number of concurrent running tasks is limited by limit, if limit is zero,
+// there is no limit.
+// The group can be stopped by calling Group.Stop which will close the group,
+// it'll not accept new tasks, and it will wait for all running tasks to
+// complete or stopTimeout to elapse, what ever comes first.
+// log is used to log any error returned by the tasks or internal errors. The
+// logs will be prefixed by errPrefix.
+func NewGroup(limit uint64, stopTimeout time.Duration, log Logger, errPrefix string) *Group {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	var logErr = func(error) {}
+	if log != nil {
+		logErr = func(err error) {
+			var format string
+			if errPrefix == "" {
+				format = "%v"
+			} else {
+				format = errPrefix + ": %v"
+			}
+			log.Errorf(format, err)
+		}
+	}
```

Comment:
Consider extracting the initialization of `logErr` into it's own helper function.  That way you can more easily unit test it, which is probably a good idea since there are four paths through it.